packages/opentelemetry-plugin-express/src/express.ts

```diff
@@ -204,12 +204,36 @@ export class ExpressPlugin extends BasePlugin<typeof express> {
             return patchEnd(span, plugin._tracer.bind(next))();
           };
         }
+        /**
+         * Ending a layer's span at the right time is quite difficult because
+         * of their nature: they can be either sync or async and either it call
+         * their callback or not, and we can't know that before calling it.
+         *
+         * Here are the different state that can result from calling the layer code:
+         *  1. The layer is sync and call its callback
+         *      - it will be correctly closed by the above code
+         *  2. The layer is sync and never call its callback
+         *      - we need to end the span manually after calling the layer code
+         *  3. The layer is async and call its callback
+         *      - it will be correctly closed by the above code
+         *  4. The layer is async and never call its callback
+         *      - we need to end the span manually
+         *      - however it will not represent the actual time that the span took
+         *      - we would need to hook into the http response's `end` event.
+         *
+         * The fix for case 4) is to hook into the `end` event, but it would apply
+         * to case 2) and 3), which would cost quite a lot of memory for big express apps.
+         *
+         * The workaround is to schedule a callback to run on next tick and check if the
+         * callback has been called. If not we close the span because we assume being
+         * in the case 2) or 4).
+         */
         const result = original.apply(this, arguments);
-        // if the layer return a response, the callback will never
-        // be called, so we need to manually close the span
-        if (callbackHasBeenCalled === false) {
-          span.end();
-        }
+        setImmediate(() => {
```

Comment:
I would save the end time before it is called and then use it to end span. WDYT ?

Reply:
I'm not sure which end time you are speaking about here, could you give an example ?

Reply:
```
const time = hrTime();
if (callbackHasBeenCalled === false) {
  span.end(time);
}
```

Reply:
```
const time = hrTime();
setImmediate(() => {
  if (callbackHasBeenCalled === false) {
    span.end(time);
  }
})

```

Reply:
That is the time it starts, not the time it ends. In any case, the difference would be less than 1ms so it doesn't really matter.

Reply:
Indeed that would be more correct, fixed it !

Reply:
Since its directly after the function call and we assume the code was synchrone i believe thats the end time